input-profile.js

```diff
@@ -0,0 +1,249 @@
+                        console.log('Loaded ' + this.handedness + '  controller');
+                    })
+                    .catch(console.error);
+            }
+        }
+    }
+
+    getGamepadObjectsFromProfile(profile, obj) {
+        //initialise components
+        const components = profile.layouts[this.handedness].components;
+        if (!components) return;
+        for (const i in components) {
+            if (components.hasOwnProperty(i)) {
+                const visualResponses = components[i].visualResponses;
+                for (const j in visualResponses) {
+                    if (visualResponses.hasOwnProperty(j)) {
+                        this.getObjectByName(obj, visualResponses[j].valueNodeName);
+                        this.getObjectByName(obj, visualResponses[j].minNodeName);
+                        this.getObjectByName(obj, visualResponses[j].maxNodeName);
+                    }
+                }
+            }
+        }
+    }
+
+    getObjectByName(obj, name) {
+        if (!obj || !name) return;
+        if (obj.name === name) this._gamepadObjects[name] = obj;
+        const children = obj.children;
+        for (let i = 0; i < children.length; ++i) this.getObjectByName(children[i], name);
+    }
+
+    assignTransform(target, min, max, value) {
+        vec3.lerp(
+            _TempVec,
+            min.getTranslationWorld([]),
+            max.getTranslationWorld([]),
+            value
+        );
+        target.setTranslationWorld(_TempVec);
+
+        quat.lerp(_TempQuat, min.rotationWorld, max.rotationWorld, value);
+        quat.normalize(_TempQuat, _TempQuat);
+        target.rotationWorld = _TempQuat;
+    }
+
+    mapGamepadInput() {
+        const components = this.ProfileJSON.layouts[this.handedness].components;
+        if (!components) return;
+        for (const i in components) {
+            if (components.hasOwnProperty(i)) {
+                const component = components[i];
+                const visualResponses = component.visualResponses;
+                for (const j in visualResponses) {
+                    if (visualResponses.hasOwnProperty) {
+                        const visualResponse = visualResponses[j];
+                        const target = this._gamepadObjects[visualResponse.valueNodeName];
+                        const min = this._gamepadObjects[visualResponse.minNodeName];
+                        const max = this._gamepadObjects[visualResponse.maxNodeName];
+
+                        this.assignTransform(
+                            target,
+                            min,
+                            max,
+                            this.getGamepadValue(component, visualResponse)
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    getGamepadValue(component, visualResponse) {
+        if (visualResponse.valueNodeProperty === 'transform') {
+            switch (component.type) {
```

Comment:
Based on my above comment about efficiency, I think having a custom intermediary data state would be much better